python/dolfinx/wrappers/graph.cpp

```diff
@@ -35,65 +37,78 @@ namespace dolfinx_wrappers
 {
 
 template <typename T>
-void declare_adjacency_list(py::module& m, std::string type)
+void declare_adjacency_list(nb::module_& m, std::string type)
 {
   std::string pyclass_name = std::string("AdjacencyList_") + type;
-  py::class_<dolfinx::graph::AdjacencyList<T>,
-             std::shared_ptr<dolfinx::graph::AdjacencyList<T>>>(
-      m, pyclass_name.c_str(), "Adjacency List")
-      .def(py::init(
-               [](const py::array_t<T, py::array::c_style>& adj)
-               {
-                 if (adj.ndim() > 2)
-                   throw std::runtime_error("Incorrect array dimension.");
-                 const std::size_t dim = adj.ndim() < 2 ? 1 : adj.shape(1);
-                 std::vector<T> data(adj.data(), adj.data() + adj.size());
-                 return dolfinx::graph::regular_adjacency_list(std::move(data),
-                                                               dim);
-               }),
-           py::arg("adj"))
-      .def(py::init(
-               [](const py::array_t<T, py::array::c_style>& array,
-                  const py::array_t<std::int32_t, py::array::c_style>& displ)
-               {
-                 std::vector<T> data(array.data(), array.data() + array.size());
-                 std::vector<std::int32_t> offsets(displ.data(),
-                                                   displ.data() + displ.size());
-                 return dolfinx::graph::AdjacencyList<T>(std::move(data),
-                                                         std::move(offsets));
-               }),
-           py::arg("data"), py::arg("offsets"))
+  nb::class_<dolfinx::graph::AdjacencyList<T>>(m, pyclass_name.c_str(),
+                                               "Adjacency List")
+      .def(
+          "__init__",
+          [](dolfinx::graph::AdjacencyList<T>* a,
+             nb::ndarray<const T, nb::ndim<1>, nb::c_contig> adj)
+          {
+            std::vector<T> data(adj.data(), adj.data() + adj.size());
+            new (a) dolfinx::graph::AdjacencyList<T>(
+                dolfinx::graph::regular_adjacency_list(std::move(data), 1));
+          },
+          nb::arg("adj").noconvert())
+      .def(
+          "__init__",
+          [](dolfinx::graph::AdjacencyList<T>* a,
+             nb::ndarray<const T, nb::ndim<2>, nb::c_contig> adj)
+          {
+            std::vector<T> data(adj.data(), adj.data() + adj.size());
+            new (a) dolfinx::graph::AdjacencyList<T>(
```

Comment:
Is using the `new` keyword necessary? The old approach looked cleaner.

Reply:
Yes. `nanobind` uses 'new placement' for construction.